File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3081_replace-question-marks-in-string-to-minimize-its-value/solutions/python3/solution.py
class Solution:
    def minimizeStringValue(self, s: str) -> str:
        from collections import Counter
        
        # We will replace the resulting array of characters
        res = list(s)
        # Keep count of occurrences of each character before each position
        count = Counter()
        
        for i in range(len(s)):
            if s[i] == '?':
                # Find the lexicographically smallest character that minimizes increase in cost
                # We choose the character that has appeared the least times to minimize the cost
                best_char = 'a'
                while best_char in count and count[best_char] > 0:
                    best_char = chr(ord(best_char) + 1)
                res[i] = best_char
                count[best_char] += 1
            else:
                # Use the existing character and update the count
                count[s[i]] += 1

        return ''.join(res)

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3081_replace-question-marks-in-string-to-minimize-its-value/solutions/cpp/solution.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    string minimizeStringValue(string s) {
        vector<int> freq(26, 0);
        // Fill frequency for known characters
        for (char c : s) {
            if (c != '?') {
                freq[c - 'a']++;
            }
        }

        // Replace '?' with the lexicographically smallest character that minimizes the value
        for (char& c : s) {
            if (c == '?') {
                for (int i = 0; i < 26; i++) {
                    if (freq[i] == 0) { // Find the smallest unused character
                        c = 'a' + i;
                        freq[i]++;
                        break;
                    }
                }
            }
        }

        return s;
    }
};
